codegen/src/SchemaProvider.scala

```diff
@@ -2,89 +2,149 @@ package besom.codegen
-    val enumTypeTokensBuffer   = ListBuffer.empty[String]
-    val objectTypeTokensBuffer = ListBuffer.empty[String]
+  def dependencies(schemaName: SchemaName, packageVersion: PackageVersion): List[(SchemaName, PackageVersion)]
 
-    // post-process the package to improve its quality
-    pulumiPackage.types.foreach {
-      case (typeToken, _: EnumTypeDefinition) =>
-        enumTypeTokensBuffer += typeToken.toLowerCase // Unifying to lower case to circumvent inconsistencies in low quality schemas (e.g. aws)
-      case (typeToken, _: ObjectTypeDefinition) =>
-        objectTypeTokensBuffer += typeToken.toLowerCase
-    }
-
-    PulumiPackageInfo(
-      enumTypeTokens = enumTypeTokensBuffer.toSet,
-      objectTypeTokens = objectTypeTokensBuffer.toSet,
-      resourceTypeTokens = pulumiPackage.resources.keySet.map(_.toLowerCase),
-      moduleToPackageParts = pulumiPackage.moduleToPackageParts,
-      providerToPackageParts = pulumiPackage.providerToPackageParts
-    )
-  }
+  def packageInfo(schemaOrMetadata: Either[SchemaFile, PackageMetadata]): (PulumiPackage, PulumiPackageInfo)
+  def packageInfo(metadata: PackageMetadata): (PulumiPackage, PulumiPackageInfo)
 }
 
 class DownloadingSchemaProvider(schemaCacheDirPath: os.Path)(implicit logger: Logger) extends SchemaProvider {
 
-  import SchemaProvider._
-
-  private val packageInfos: collection.mutable.Map[(ProviderName, SchemaVersion), PulumiPackageInfo] =
+  private val schemaFileName = "schema.json"
+  private val packageInfos: collection.mutable.Map[(SchemaName, PackageVersion), PulumiPackageInfo] =
     collection.mutable.Map.empty
 
-  private def downloadedSchemaFilePath(providerName: ProviderName, schemaVersion: SchemaVersion): os.Path = {
-    val schemaFilePath = schemaCacheDirPath / providerName / schemaVersion / "schema.json"
+  private def pulumiPackage(metadata: PackageMetadata): (PulumiPackage, PackageMetadata) = {
+    val schemaFilePath = schemaCacheDirPath / metadata.name / metadata.version / schemaFileName
 
     if (!os.exists(schemaFilePath)) {
-      logger.debug(s"Downloading schema for $providerName:$schemaVersion' into '${schemaFilePath.relativeTo(os.pwd)}'")
-      val schemaSource = s"$providerName@$schemaVersion"
-      os.makeDir.all(schemaFilePath / os.up)
-      os.proc("pulumi", "package", "get-schema", schemaSource).call(stdout = schemaFilePath)
+      logger.debug(
+        s"Downloading schema for '${metadata.name}:${metadata.version}' into '${schemaFilePath.relativeTo(os.pwd)}'"
+      )
+
+      val schemaSource =
+        if (metadata.version.isDefault)
+          metadata.name
+        else
+          s"${metadata.name}@${metadata.version}"
+
+      val installCmd: List[String] =
+        List("pulumi", "plugin", "install", "resource", metadata.name) ++ {
+          // use version only if it is not the default, otherwise try to install the latest
+          if (metadata.version.isDefault) List(metadata.version) else List.empty
```

Comment:
```suggestion
          if (!metadata.version.isDefault) List(metadata.version) else List.empty
```
Otherwise the implementation seems to be against the comment.
But if the default version is `0.0.0`, which is only used in codegen tests, without the intent to actually download this schema, maybe `installCmd` shouldn't be run for the default version at all?